File: api/access_token.go
package api

import (
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/google/uuid"
)

type AccessToken struct {
	ID         uuid.UUID           `json:"id"`
	CreatedAt  time.Time           `json:"createdAt"`
	ExpiresAt  *time.Time          `json:"expiresAt,omitempty"`
	LastUsedAt *time.Time          `json:"lastUsedAt,omitempty"`
	Label      *string             `json:"label,omitempty"`
	UserRole   *types.UserRole     `json:"userRole,omitempty"`
	Secrets    []AccessTokenSecret `json:"secrets"`
}

type AccessTokenSecret struct {
	Slot       types.AccessTokenSecretSlot `json:"slot"`
	CreatedAt  time.Time                   `json:"createdAt"`
	LastUsedAt *time.Time                  `json:"lastUsedAt,omitempty"`
}

func (obj AccessToken) WithKey(key string) AccessTokenWithKey {
	return AccessTokenWithKey{obj, key}
}

type AccessTokenWithKey struct {
	AccessToken
	Key string `json:"key"`
}

type CreateAccessTokenRequest struct {
	ExpiresAt *time.Time      `json:"expiresAt"`
	Label     *string         `json:"label"`
	UserRole  *types.UserRole `json:"userRole"`
}

File: sdk/js/src/types/access-token.ts
import {BaseModel} from './base';
import {UserRole} from './user-account';

export type AccessTokenSecretSlot = 1 | 2;

export interface AccessTokenSecret {
  slot: AccessTokenSecretSlot;
  createdAt: string;
  lastUsedAt?: string;
}

export interface AccessToken extends BaseModel {
  expiresAt?: string;
  lastUsedAt?: string;
  label?: string;
  userRole?: UserRole;
  secrets: AccessTokenSecret[];
}

export interface AccessTokenWithKey extends AccessToken {
  key: string;
}

export interface CreateAccessTokenRequest {
  label?: string;
  expiresAt?: Date;
  userRole?: UserRole;
}
